From 09d82602339406fdacd593269886668cc20df651 Mon Sep 17 00:00:00 2001
From: Rob Reynolds <ferventcoder@gmail.com>
Date: Thu, 3 Jul 2014 16:19:10 -0500
Subject: [PATCH] Get description failures are not exceptional

 - This adds an additional `ERROR_RESOURCE_NAME_NOT_FOUND` (1814) to
   the items to not throw an error on in get_config2_info.
 - This also adds the approach that an error getting description, for whatever
   reason, is not an exceptional behavior and therefore should not throw an
   error.
---
 lib/win32/service.rb           | 18 ++++++++++++------
 lib/win32/windows/constants.rb |  1 +
 2 files changed, 13 insertions(+), 6 deletions(-)

diff --git a/lib/win32/service.rb b/lib/win32/service.rb
index 6085f35..b3fd6b6 100644
--- a/lib/win32/service.rb
+++ b/lib/win32/service.rb
@@ -1075,12 +1075,18 @@ module Win32
 
               deps = config_struct[:lpDependencies].read_array_of_null_separated_strings
 
-              buf = get_config2_info(handle_scs, SERVICE_CONFIG_DESCRIPTION)
-
-              if buf.is_a?(Fixnum) || buf.read_pointer.null?
+              begin
+                buf = get_config2_info(handle_scs, SERVICE_CONFIG_DESCRIPTION)
+
+                if buf.is_a?(Fixnum) || buf.read_pointer.null?
+                  description = ''
+                else
+                  description = buf.read_pointer.read_string
+                end
+              rescue
+                # While being annoying, not being able to get a description is not exceptional
+                warn "WARNING: Failed to retreive description for the #{service_name} service."
                 description = ''
-              else
-                description = buf.read_pointer.read_string
               end
 
               delayed_start_buf = get_config2_info(handle_scs, SERVICE_CONFIG_DELAYED_AUTO_START_INFO)
@@ -1367,7 +1373,7 @@ module Win32
       #
       if !bool && err_num == ERROR_INSUFFICIENT_BUFFER
         config2_buf = FFI::MemoryPointer.new(:char, bytes_needed.read_ulong)
-      elsif [ERROR_FILE_NOT_FOUND, ERROR_RESOURCE_TYPE_NOT_FOUND].include?(err_num)
+      elsif [ERROR_FILE_NOT_FOUND, ERROR_RESOURCE_TYPE_NOT_FOUND, ERROR_RESOURCE_NAME_NOT_FOUND].include?(err_num)
         return err_num
       else
         CloseServiceHandle(handle)
diff --git a/lib/win32/windows/constants.rb b/lib/win32/windows/constants.rb
index 21ff3ea..b6e753d 100644
--- a/lib/win32/windows/constants.rb
+++ b/lib/win32/windows/constants.rb
@@ -137,6 +137,7 @@ module Windows
     ERROR_MORE_DATA = 234
     ERROR_FILE_NOT_FOUND = 2
     ERROR_RESOURCE_TYPE_NOT_FOUND = 1813
+    ERROR_RESOURCE_NAME_NOT_FOUND = 1814
     WAIT_FAILED = 0xFFFFFFFF
   end
 end
-- 
1.9.4.msysgit.0

